tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11299,6 +11299,62 @@ bool Sema::DiagnoseHLSLMethodCall(const CXXMethodDecl *MD, SourceLocation Loc) {
   return false;
 }
 
+void ValidateWaveSize(clang::Sema *S, FunctionDecl *FD) {
+  const auto *SM =
+      hlsl::ShaderModel::GetByName(S->getLangOpts().HLSLProfile.c_str());
+
+  const llvm::StringRef &entryName = S->getLangOpts().HLSLEntryFunction;
+  StringRef functionName = "";
+  if (FD->getIdentifier())
+    functionName = FD->getName();
+  bool isEntry = false;
+  bool isNode = false;
+  bool isCS = false;
+
+  isEntry =
+      !SM->IsLib() &&
+      FD->getDeclContext()->getDeclKind() == Decl::Kind::TranslationUnit &&
+      functionName == entryName;
+
+  auto *shaderAttribute = FD->getAttr<HLSLShaderAttr>();
+  StringRef Literal = shaderAttribute->getStage();
+  DXIL::ShaderKind Stage = ShaderModel::KindFromFullName(Literal);
+  isNode = Stage == DXIL::ShaderKind::Node;
+  isCS = Stage == DXIL::ShaderKind::Compute;
+
+  HLSLWaveSizeAttr *attr = FD->getAttr<HLSLWaveSizeAttr>();
+
+  if (!SM->IsLib() && !isEntry && !isNode) {
```

Comment:
This is the only use of isEntry. As a result, isEntry can be simplified to exclude the lib check. 

The effect of this and the code above is this check (lightly edited for brevity):

`if (!isLIb && (isLib || DeclKind != TranslationUnit || functionname != entryname) && !isNode)`

Since the only way to even get through short-circuiting to the second clause is for this to not be a library, the check for the library there isn't needed. 

I'm torn about the final !isNode clause. At present, the only way to have a node is to be compiling as a library, but I'm not quite confident enough about that always being the case to suggest removing it.